### ISSUE TITLE
Upper/lower bounds

### ISSUE THREAD

**Issue by mindplay-dk** [mindplay-dk/php-generics-rfc#3]
The RFC currently describes upper bounds only.

Decide whether to support both upper and lower bounds.

Decide on the syntax.

> Scala spells an upper bound as `T <: A`. A lower bound is `T >: A`.
>
> C# spells an upper bound as `T: A`. It does not support lower bounds directly. In some cases there can be two type parameters and you can specify one is the upper bound for the other, effectively making the reverse a lower-bound:
>
> `void F<T,U> (T source, U dest) where T : U`

Add/update relevant sections of the RFC as needed.

**Comment by mindplay-dk** [comment]
Note that the "Generic Types" section currently doesn't mention lower bounds - if we support both upper and lower bounds, will the `is` syntax suffice, or do we need something different?

**Comment by morrisonlevi** [comment]
Personally I want to see more examples of lower-bounds in other languages to help make this decision.

**Comment by natebrunette** [comment]
I don't think we need to support lower bounds if we're going to support declaration-site variance. I believe that you would only use a lower bound for `in` parameters, and specifying a parameter as `in` should be sufficient.

**Comment by natebrunette** [comment]
We would also need to support these keywords in type hints as well.

For example, consider the following

```php
interface List<E>
{
    public function get(int $index): E;
    public function add(E $element): bool;
}

function addStringToList(List<in string> $writeTo, string $value) {}
```

Type parameter `E` is invariant, but the method `addStringToList()` would accept a list of type `List<string>` or `List<mixed>`.

**Comment by morrisonlevi** [comment]
>  I believe that you would only use a lower bound for in parameters, and specifying a parameter as in should be sufficient.

This could be true but I'd like to see more examples in other languages.

> We would also need to support these keywords in type hints as well.

Does C# allow this?

**Comment by morrisonlevi** [comment]
Let me try to make sense of what you are saying. The `E` in `List<E>` is invariant, precisely because it is used in both input and output positions. Then when you write `List<in string>` you want it to somehow make `get()` be inaccessible or an error to call?

I don't think this would work. We won't have the necessary type information at compile time to forbid the call there, and at runtime the method has whatever methods it has.

**Comment by natebrunette** [comment]
We don't allow this, right? What are your current thoughts on how to forbid it?

```php
interface ReadonlyList<out E>
{
    public function get(int $index): E;
    public function add(E $element): bool;
}
```

Do we not know which methods are called until runtime?
